src/utils/format_open_telemetry/format_open_telemetry.cc

```diff
@@ -0,0 +1,168 @@
+
+#include "opentelemetry/proto/collector/metrics/v1/metrics_service.pb.h"
+#include "opentelemetry/proto/common/v1/common.pb.h"
+#include "opentelemetry/proto/metrics/v1/metrics.pb.h"
+#include "opentelemetry/proto/resource/v1/resource.pb.h"
+
+using opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest;
+using opentelemetry::proto::common::v1::AnyValue;
+using opentelemetry::proto::common::v1::InstrumentationScope;
+using opentelemetry::proto::common::v1::KeyValue;
+using opentelemetry::proto::metrics::v1::AGGREGATION_TEMPORALITY_CUMULATIVE;
+using opentelemetry::proto::metrics::v1::Gauge;
+using opentelemetry::proto::metrics::v1::Metric;
+using opentelemetry::proto::metrics::v1::NumberDataPoint;
+using opentelemetry::proto::metrics::v1::ResourceMetrics;
+using opentelemetry::proto::metrics::v1::ScopeMetrics;
+using opentelemetry::proto::metrics::v1::Sum;
+using opentelemetry::proto::resource::v1::Resource;
+
+static void metric_to_number_data_point(NumberDataPoint *dp,
+                                        metric_t const *m) {
+  for (size_t i = 0; i < m->label.num; i++) {
+    label_pair_t *l = m->label.ptr + i;
+
+    KeyValue *kv = dp->add_attributes();
+    kv->set_key(l->name);
+    AnyValue *v = kv->mutable_value();
+    v->set_string_value(l->value);
+  }
+
+  dp->set_time_unix_nano(CDTIME_T_TO_NS(m->time));
+  // TODO(): also set "start time". We may need to use the cache to determine
+  // when we've seen a metric for the first time.
+
+  switch (m->family->type) {
+  case METRIC_TYPE_COUNTER:
+    dp->set_as_int(m->value.derive);
+    break;
+  case METRIC_TYPE_GAUGE:
+    dp->set_as_double(m->value.gauge);
+    break;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
+  }
+}
+
+static void set_sum(Metric *m, metric_family_t const *fam) {
+  Sum *s = m->mutable_sum();
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    metric_t const *m = fam->metric.ptr + i;
+    assert(m->family == fam);
+
+    NumberDataPoint *dp = s->add_data_points();
+    metric_to_number_data_point(dp, fam->metric.ptr + i);
+  }
+
+  s->set_aggregation_temporality(AGGREGATION_TEMPORALITY_CUMULATIVE);
+  s->set_is_monotonic(true);
+}
+
+static void set_gauge(Metric *m, metric_family_t const *fam) {
+  Gauge *g = m->mutable_gauge();
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    NumberDataPoint *dp = g->add_data_points();
+    metric_to_number_data_point(dp, fam->metric.ptr + i);
+  }
+}
+
+static void add_metric(ScopeMetrics *sm, metric_family_t const *fam) {
+  Metric *m = sm->add_metrics();
+
+  m->set_name(fam->name);
+  if (fam->help != NULL) {
+    m->set_description(fam->help);
+  }
+  if (fam->unit != NULL) {
+    m->set_unit(fam->unit);
+  }
+
+  switch (fam->type) {
+  case METRIC_TYPE_COUNTER:
+    set_sum(m, fam);
+    return;
+  case METRIC_TYPE_GAUGE:
+    set_gauge(m, fam);
+    return;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
```

Comment:
`FATAL()`?

Reply:
As above, I have implemented graceful handling of this situation. The case must remain, though, otherwise compilers complain about an unhandled value of the enum.